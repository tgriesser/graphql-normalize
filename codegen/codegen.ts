import type { CodegenPlugin } from '@graphql-codegen/plugin-helpers';

const NormalizeCodegenPlugin: CodegenPlugin = {
  plugin(schema, documents, config, info) {
    //
  },
};

export default NormalizeCodegenPlugin;
