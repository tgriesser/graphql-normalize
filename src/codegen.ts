import type { CodegenPlugin } from '@graphql-codegen/plugin-helpers';

export { generateNormalizedMetadata } from './codegen/generateNormalizedMetadata.js';
export { generateNormalizedOperation } from './codegen/generateNormalizedOperation.js';

const NormalizeCodegenPlugin: CodegenPlugin = {
  plugin(schema, documents, config, info) {
    return '';
  },
};

export default NormalizeCodegenPlugin;
