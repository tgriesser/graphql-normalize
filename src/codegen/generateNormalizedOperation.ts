import { optimizeDocuments } from '@graphql-tools/relay-operation-optimizer';
import { DocumentNode, GraphQLSchema, Kind } from 'graphql';
import { addNormalizingFields } from './addNormalizingFields';
import type { TypePolicies } from './getCacheKey';

export function generateNormalizedOperation(
  schema: GraphQLSchema,
  document: DocumentNode,
  typePolicies: TypePolicies = {}
) {
  return generateNormalizedOperations(schema, [document], typePolicies)[0];
}

/**
 * Given a schema & operation, flattens the fragments and adds any necessary missing fields
 * in order to properly normalize the response in a cache.
 */
export function generateNormalizedOperations(
  schema: GraphQLSchema,
  documents: DocumentNode[],
  typePolicies: TypePolicies = {}
) {
  let toNormalize = documents;
  if (documents.some((o) => o.definitions.some((d) => d.kind === Kind.FRAGMENT_DEFINITION))) {
    toNormalize = optimizeDocuments(schema, documents, {
      noLocation: true,
      includeFragments: false,
    });
  }
  return addNormalizingFields(schema, toNormalize, typePolicies);
}
