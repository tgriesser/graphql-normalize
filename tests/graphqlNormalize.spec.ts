import { ExecutionResult, execute } from 'graphql';
import { beforeEach, describe, expect, it } from 'vitest';
import _ from 'lodash';

import { generateNormalizedOperation } from '../src/codegen/generateNormalizedOperation';
import { schema } from './fixtures/schema';
import { operation1Doc } from './fixtures/ops';
import { generateNormalizedMetadata } from '../src/codegen/generateNormalizedMetadata';
import { graphqlNormalize } from '../src/graphqlNormalize';
import type { NormalizeMetaShape } from '../src/metadataShapes';
import type { CacheShape } from '../src/cache';
import { enablePatches, produceWithPatches } from 'immer';

enablePatches();

describe('syncWithCache', () => {
  let meta: NormalizeMetaShape;
  let variableValues = {};
  let cache: CacheShape;
  let result: ExecutionResult;

  beforeEach(async () => {
    meta = generateNormalizedMetadata(schema, operation1Doc);
    variableValues = {};
    result = await execute({
      schema,
      variableValues,
      document: generateNormalizedOperation(schema, operation1Doc),
    });
    cache = {
      operations: {},
      fields: {},
    } as const;
  });

  it('syncs the query result with the cache', async () => {
    const obj = graphqlNormalize({
      action: 'write',
      meta,
      variableValues,
      operationResult: result,
      cache,
      isEqual: _.isEqual,
      //
    });

    expect(cache).toMatchSnapshot();
    expect(obj).toMatchSnapshot();

    const obj2 = graphqlNormalize({
      action: 'write',
      meta,
      variableValues,
      operationResult: result,
      currentResult: obj.result,
      cache,
      isEqual: _.isEqual,
    });

    expect({
      added: obj2.added,
      updated: obj2.updated,
    }).toMatchInlineSnapshot(`
      {
        "added": 0,
        "updated": 0,
      }
    `);
  });

  it('writes additional values into the store when the variables are changed', async () => {
    const { result: currentResult } = graphqlNormalize({
      action: 'write',
      meta,
      variableValues,
      operationResult: result,
      cache,
      isEqual: _.isEqual,
      //
    });

    const variableValues2 = { hasNode: true, nodeId: 'VXNlcjox' };

    const [, patches] = await produceWithPatches({ cache, currentResult }, async ({ cache, currentResult }) => {
      const sync2 = graphqlNormalize({
        action: 'write',
        meta,
        variableValues: variableValues2,
        currentResult,
        operationResult: await execute({
          schema,
          variableValues: variableValues2,
          contextValue: {},
          document: generateNormalizedOperation(schema, operation1Doc),
        }),
        cache,
        isEqual: _.isEqual,
      });

      // Only added a single item to the cache, since we already have this object.
      expect(sync2.added).toEqual(1);
      expect(sync2.updated).toEqual(0);
    });

    expect(patches).toMatchSnapshot();
  });

  it('removes unused keys when keys dissapear', async () => {
    const variableValues = { hasNode: true, nodeId: 'VXNlcjox' };

    const { result: currentResult } = graphqlNormalize({
      action: 'write',
      meta,
      variableValues,
      operationResult: await execute({
        schema,
        variableValues,
        contextValue: {},
        document: generateNormalizedOperation(schema, operation1Doc),
      }),
      cache,
      isEqual: _.isEqual,
      //
    });

    const [, patches] = await produceWithPatches({ cache, currentResult }, async ({ cache, currentResult }) => {
      graphqlNormalize({
        action: 'write',
        meta,
        variableValues: {},
        currentResult,
        operationResult: await execute({
          schema,
          variableValues,
          contextValue: {},
          document: generateNormalizedOperation(schema, operation1Doc),
        }),
        cache,
        isEqual: _.isEqual,
      });
    });

    expect(patches).toMatchSnapshot();
  });
});
