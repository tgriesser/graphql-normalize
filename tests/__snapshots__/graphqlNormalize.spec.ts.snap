// Vitest Snapshot v1

exports[`syncWithCache > removes unused keys when keys dissapear 1`] = `
[
  {
    "op": "remove",
    "path": [
      "currentResult",
      "currentNode",
    ],
  },
]
`;

exports[`syncWithCache > syncs the query result with the cache 1`] = `
{
  "Blog:QmxvZzox": {
    "__typename": "Blog",
    "id": {
      "$": "QmxvZzox",
    },
    "name": {
      "$": "My Cool #1",
    },
  },
  "Blog:QmxvZzoy": {
    "__typename": "Blog",
    "id": {
      "$": "QmxvZzoy",
    },
    "name": {
      "$": "My Cool #2",
    },
  },
  "Blog:QmxvZzoz": {
    "__typename": "Blog",
    "id": {
      "$": "QmxvZzoz",
    },
    "name": {
      "$": "My Cool #3",
    },
  },
  "Post:UG9zdDo0": {
    "__typename": "Post",
    "author": {
      "$": {
        "$ref": "User:VXNlcjoy",
      },
    },
    "id": {
      "$": "UG9zdDo0",
    },
    "title": {
      "$": "Some Blog post 4",
    },
  },
  "Post:UG9zdDo1": {
    "__typename": "Post",
    "author": {
      "$": {
        "$ref": "User:VXNlcjox",
      },
    },
    "id": {
      "$": "UG9zdDo1",
    },
    "title": {
      "$": "Some Blog post 5",
    },
  },
  "Post:UG9zdDo2": {
    "__typename": "Post",
    "author": {
      "$": {
        "$ref": "User:VXNlcjoy",
      },
    },
    "id": {
      "$": "UG9zdDo2",
    },
    "title": {
      "$": "Some Blog post 6",
    },
  },
  "Post:UG9zdDo3": {
    "__typename": "Post",
    "author": {
      "$": {
        "$ref": "User:VXNlcjox",
      },
    },
    "id": {
      "$": "UG9zdDo3",
    },
    "title": {
      "$": "Some Blog post 7",
    },
  },
  "Post:UG9zdDo4": {
    "__typename": "Post",
    "author": {
      "$": {
        "$ref": "User:VXNlcjoy",
      },
    },
    "id": {
      "$": "UG9zdDo4",
    },
    "title": {
      "$": "Some Blog post 8",
    },
  },
  "Post:UG9zdDo5": {
    "__typename": "Post",
    "author": {
      "$": {
        "$ref": "User:VXNlcjox",
      },
    },
    "id": {
      "$": "UG9zdDo5",
    },
    "title": {
      "$": "Some Blog post 9",
    },
  },
  "Post:UG9zdDox": {
    "__typename": "Post",
    "author": {
      "$": {
        "$ref": "User:VXNlcjox",
      },
    },
    "id": {
      "$": "UG9zdDox",
    },
    "title": {
      "$": "Some Blog post 1",
    },
  },
  "Post:UG9zdDoxMA==": {
    "__typename": "Post",
    "author": {
      "$": {
        "$ref": "User:VXNlcjoy",
      },
    },
    "id": {
      "$": "UG9zdDoxMA==",
    },
    "title": {
      "$": "Some Blog post 10",
    },
  },
  "Post:UG9zdDoxMQ==": {
    "__typename": "Post",
    "author": {
      "$": {
        "$ref": "User:VXNlcjox",
      },
    },
    "id": {
      "$": "UG9zdDoxMQ==",
    },
  },
  "Post:UG9zdDoxMg==": {
    "__typename": "Post",
    "author": {
      "$": {
        "$ref": "User:VXNlcjoy",
      },
    },
    "id": {
      "$": "UG9zdDoxMg==",
    },
  },
  "Post:UG9zdDoy": {
    "__typename": "Post",
    "author": {
      "$": {
        "$ref": "User:VXNlcjoy",
      },
    },
    "id": {
      "$": "UG9zdDoy",
    },
    "title": {
      "$": "Some Blog post 2",
    },
  },
  "Post:UG9zdDoz": {
    "__typename": "Post",
    "author": {
      "$": {
        "$ref": "User:VXNlcjox",
      },
    },
    "id": {
      "$": "UG9zdDoz",
    },
    "title": {
      "$": "Some Blog post 3",
    },
  },
  "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjo0": {
    "__typename": "QueryPostsConnectionEdge",
    "node": {
      "$": {
        "$ref": "Post:UG9zdDo1",
      },
    },
  },
  "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjo1": {
    "__typename": "QueryPostsConnectionEdge",
    "node": {
      "$": {
        "$ref": "Post:UG9zdDo2",
      },
    },
  },
  "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjo2": {
    "__typename": "QueryPostsConnectionEdge",
    "node": {
      "$": {
        "$ref": "Post:UG9zdDo3",
      },
    },
  },
  "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjo3": {
    "__typename": "QueryPostsConnectionEdge",
    "node": {
      "$": {
        "$ref": "Post:UG9zdDo4",
      },
    },
  },
  "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjo4": {
    "__typename": "QueryPostsConnectionEdge",
    "node": {
      "$": {
        "$ref": "Post:UG9zdDo5",
      },
    },
  },
  "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjo5": {
    "__typename": "QueryPostsConnectionEdge",
    "node": {
      "$": {
        "$ref": "Post:UG9zdDoxMA==",
      },
    },
  },
  "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjow": {
    "__typename": "QueryPostsConnectionEdge",
    "node": {
      "$": {
        "$ref": "Post:UG9zdDox",
      },
    },
  },
  "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjox": {
    "__typename": "QueryPostsConnectionEdge",
    "node": {
      "$": {
        "$ref": "Post:UG9zdDoy",
      },
    },
  },
  "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjoxMA==": {
    "__typename": "QueryPostsConnectionEdge",
    "node": {
      "$": {
        "$ref": "Post:UG9zdDoxMQ==",
      },
    },
  },
  "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjoxMQ==": {
    "__typename": "QueryPostsConnectionEdge",
    "node": {
      "$": {
        "$ref": "Post:UG9zdDoxMg==",
      },
    },
  },
  "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjoy": {
    "__typename": "QueryPostsConnectionEdge",
    "node": {
      "$": {
        "$ref": "Post:UG9zdDoz",
      },
    },
  },
  "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjoz": {
    "__typename": "QueryPostsConnectionEdge",
    "node": {
      "$": {
        "$ref": "Post:UG9zdDo0",
      },
    },
  },
  "User:VXNlcjox": {
    "__typename": "User",
    "id": {
      "$": "VXNlcjox",
    },
    "name": {
      "$": "user 1",
    },
  },
  "User:VXNlcjoy": {
    "__typename": "User",
    "id": {
      "$": "VXNlcjoy",
    },
    "name": {
      "$": "user 2",
    },
  },
  "homeItems": {
    "{\\"first\\":\\"10\\"}": {
      "nodes": {
        "$": [
          {
            "__typename": "Comment",
          },
          {
            "$ref": "Blog:QmxvZzox",
          },
          {
            "$ref": "Post:UG9zdDox",
          },
          {
            "__typename": "User",
          },
          {
            "__typename": "Comment",
          },
          {
            "$ref": "Blog:QmxvZzoy",
          },
          {
            "$ref": "Post:UG9zdDoy",
          },
          {
            "__typename": "User",
          },
          {
            "__typename": "Comment",
          },
          {
            "$ref": "Blog:QmxvZzoz",
          },
        ],
      },
    },
  },
  "nestedString": {
    "$": [
      [
        "a",
        "b",
      ],
      [
        "foo",
      ],
    ],
  },
  "nestedUserList": {
    "$": [
      [
        {
          "$ref": "User:VXNlcjox",
        },
      ],
      [
        {
          "$ref": "User:VXNlcjoy",
        },
      ],
    ],
  },
  "num": {
    "$": 1,
  },
  "posts": {
    "{\\"first\\":\\"10\\"}": {
      "nodes": {
        "$": [
          {
            "$ref": "Post:UG9zdDox",
          },
          {
            "$ref": "Post:UG9zdDoy",
          },
          {
            "$ref": "Post:UG9zdDoz",
          },
          {
            "$ref": "Post:UG9zdDo0",
          },
          {
            "$ref": "Post:UG9zdDo1",
          },
          {
            "$ref": "Post:UG9zdDo2",
          },
          {
            "$ref": "Post:UG9zdDo3",
          },
          {
            "$ref": "Post:UG9zdDo4",
          },
          {
            "$ref": "Post:UG9zdDo5",
          },
          {
            "$ref": "Post:UG9zdDoxMA==",
          },
        ],
      },
    },
    "{\\"first\\":\\"12\\"}": {
      "edges": {
        "$": [
          {
            "$ref": "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjow",
          },
          {
            "$ref": "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjox",
          },
          {
            "$ref": "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjoy",
          },
          {
            "$ref": "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjoz",
          },
          {
            "$ref": "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjo0",
          },
          {
            "$ref": "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjo1",
          },
          {
            "$ref": "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjo2",
          },
          {
            "$ref": "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjo3",
          },
          {
            "$ref": "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjo4",
          },
          {
            "$ref": "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjo5",
          },
          {
            "$ref": "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjoxMA==",
          },
          {
            "$ref": "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjoxMQ==",
          },
        ],
      },
      "nodes": {
        "$": [
          {
            "$ref": "Post:UG9zdDox",
          },
          {
            "$ref": "Post:UG9zdDoy",
          },
          {
            "$ref": "Post:UG9zdDoz",
          },
          {
            "$ref": "Post:UG9zdDo0",
          },
          {
            "$ref": "Post:UG9zdDo1",
          },
          {
            "$ref": "Post:UG9zdDo2",
          },
          {
            "$ref": "Post:UG9zdDo3",
          },
          {
            "$ref": "Post:UG9zdDo4",
          },
          {
            "$ref": "Post:UG9zdDo5",
          },
          {
            "$ref": "Post:UG9zdDoxMA==",
          },
          {
            "$ref": "Post:UG9zdDoxMQ==",
          },
          {
            "$ref": "Post:UG9zdDoxMg==",
          },
        ],
      },
    },
  },
}
`;

exports[`syncWithCache > syncs the query result with the cache 2`] = `
{
  "added": 86,
  "cache": {
    "Blog:QmxvZzox": {
      "__typename": "Blog",
      "id": {
        "$": "QmxvZzox",
      },
      "name": {
        "$": "My Cool #1",
      },
    },
    "Blog:QmxvZzoy": {
      "__typename": "Blog",
      "id": {
        "$": "QmxvZzoy",
      },
      "name": {
        "$": "My Cool #2",
      },
    },
    "Blog:QmxvZzoz": {
      "__typename": "Blog",
      "id": {
        "$": "QmxvZzoz",
      },
      "name": {
        "$": "My Cool #3",
      },
    },
    "Post:UG9zdDo0": {
      "__typename": "Post",
      "author": {
        "$": {
          "$ref": "User:VXNlcjoy",
        },
      },
      "id": {
        "$": "UG9zdDo0",
      },
      "title": {
        "$": "Some Blog post 4",
      },
    },
    "Post:UG9zdDo1": {
      "__typename": "Post",
      "author": {
        "$": {
          "$ref": "User:VXNlcjox",
        },
      },
      "id": {
        "$": "UG9zdDo1",
      },
      "title": {
        "$": "Some Blog post 5",
      },
    },
    "Post:UG9zdDo2": {
      "__typename": "Post",
      "author": {
        "$": {
          "$ref": "User:VXNlcjoy",
        },
      },
      "id": {
        "$": "UG9zdDo2",
      },
      "title": {
        "$": "Some Blog post 6",
      },
    },
    "Post:UG9zdDo3": {
      "__typename": "Post",
      "author": {
        "$": {
          "$ref": "User:VXNlcjox",
        },
      },
      "id": {
        "$": "UG9zdDo3",
      },
      "title": {
        "$": "Some Blog post 7",
      },
    },
    "Post:UG9zdDo4": {
      "__typename": "Post",
      "author": {
        "$": {
          "$ref": "User:VXNlcjoy",
        },
      },
      "id": {
        "$": "UG9zdDo4",
      },
      "title": {
        "$": "Some Blog post 8",
      },
    },
    "Post:UG9zdDo5": {
      "__typename": "Post",
      "author": {
        "$": {
          "$ref": "User:VXNlcjox",
        },
      },
      "id": {
        "$": "UG9zdDo5",
      },
      "title": {
        "$": "Some Blog post 9",
      },
    },
    "Post:UG9zdDox": {
      "__typename": "Post",
      "author": {
        "$": {
          "$ref": "User:VXNlcjox",
        },
      },
      "id": {
        "$": "UG9zdDox",
      },
      "title": {
        "$": "Some Blog post 1",
      },
    },
    "Post:UG9zdDoxMA==": {
      "__typename": "Post",
      "author": {
        "$": {
          "$ref": "User:VXNlcjoy",
        },
      },
      "id": {
        "$": "UG9zdDoxMA==",
      },
      "title": {
        "$": "Some Blog post 10",
      },
    },
    "Post:UG9zdDoxMQ==": {
      "__typename": "Post",
      "author": {
        "$": {
          "$ref": "User:VXNlcjox",
        },
      },
      "id": {
        "$": "UG9zdDoxMQ==",
      },
    },
    "Post:UG9zdDoxMg==": {
      "__typename": "Post",
      "author": {
        "$": {
          "$ref": "User:VXNlcjoy",
        },
      },
      "id": {
        "$": "UG9zdDoxMg==",
      },
    },
    "Post:UG9zdDoy": {
      "__typename": "Post",
      "author": {
        "$": {
          "$ref": "User:VXNlcjoy",
        },
      },
      "id": {
        "$": "UG9zdDoy",
      },
      "title": {
        "$": "Some Blog post 2",
      },
    },
    "Post:UG9zdDoz": {
      "__typename": "Post",
      "author": {
        "$": {
          "$ref": "User:VXNlcjox",
        },
      },
      "id": {
        "$": "UG9zdDoz",
      },
      "title": {
        "$": "Some Blog post 3",
      },
    },
    "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjo0": {
      "__typename": "QueryPostsConnectionEdge",
      "node": {
        "$": {
          "$ref": "Post:UG9zdDo1",
        },
      },
    },
    "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjo1": {
      "__typename": "QueryPostsConnectionEdge",
      "node": {
        "$": {
          "$ref": "Post:UG9zdDo2",
        },
      },
    },
    "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjo2": {
      "__typename": "QueryPostsConnectionEdge",
      "node": {
        "$": {
          "$ref": "Post:UG9zdDo3",
        },
      },
    },
    "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjo3": {
      "__typename": "QueryPostsConnectionEdge",
      "node": {
        "$": {
          "$ref": "Post:UG9zdDo4",
        },
      },
    },
    "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjo4": {
      "__typename": "QueryPostsConnectionEdge",
      "node": {
        "$": {
          "$ref": "Post:UG9zdDo5",
        },
      },
    },
    "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjo5": {
      "__typename": "QueryPostsConnectionEdge",
      "node": {
        "$": {
          "$ref": "Post:UG9zdDoxMA==",
        },
      },
    },
    "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjow": {
      "__typename": "QueryPostsConnectionEdge",
      "node": {
        "$": {
          "$ref": "Post:UG9zdDox",
        },
      },
    },
    "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjox": {
      "__typename": "QueryPostsConnectionEdge",
      "node": {
        "$": {
          "$ref": "Post:UG9zdDoy",
        },
      },
    },
    "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjoxMA==": {
      "__typename": "QueryPostsConnectionEdge",
      "node": {
        "$": {
          "$ref": "Post:UG9zdDoxMQ==",
        },
      },
    },
    "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjoxMQ==": {
      "__typename": "QueryPostsConnectionEdge",
      "node": {
        "$": {
          "$ref": "Post:UG9zdDoxMg==",
        },
      },
    },
    "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjoy": {
      "__typename": "QueryPostsConnectionEdge",
      "node": {
        "$": {
          "$ref": "Post:UG9zdDoz",
        },
      },
    },
    "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjoz": {
      "__typename": "QueryPostsConnectionEdge",
      "node": {
        "$": {
          "$ref": "Post:UG9zdDo0",
        },
      },
    },
    "User:VXNlcjox": {
      "__typename": "User",
      "id": {
        "$": "VXNlcjox",
      },
      "name": {
        "$": "user 1",
      },
    },
    "User:VXNlcjoy": {
      "__typename": "User",
      "id": {
        "$": "VXNlcjoy",
      },
      "name": {
        "$": "user 2",
      },
    },
    "homeItems": {
      "{\\"first\\":\\"10\\"}": {
        "nodes": {
          "$": [
            {
              "__typename": "Comment",
            },
            {
              "$ref": "Blog:QmxvZzox",
            },
            {
              "$ref": "Post:UG9zdDox",
            },
            {
              "__typename": "User",
            },
            {
              "__typename": "Comment",
            },
            {
              "$ref": "Blog:QmxvZzoy",
            },
            {
              "$ref": "Post:UG9zdDoy",
            },
            {
              "__typename": "User",
            },
            {
              "__typename": "Comment",
            },
            {
              "$ref": "Blog:QmxvZzoz",
            },
          ],
        },
      },
    },
    "nestedString": {
      "$": [
        [
          "a",
          "b",
        ],
        [
          "foo",
        ],
      ],
    },
    "nestedUserList": {
      "$": [
        [
          {
            "$ref": "User:VXNlcjox",
          },
        ],
        [
          {
            "$ref": "User:VXNlcjoy",
          },
        ],
      ],
    },
    "num": {
      "$": 1,
    },
    "posts": {
      "{\\"first\\":\\"10\\"}": {
        "nodes": {
          "$": [
            {
              "$ref": "Post:UG9zdDox",
            },
            {
              "$ref": "Post:UG9zdDoy",
            },
            {
              "$ref": "Post:UG9zdDoz",
            },
            {
              "$ref": "Post:UG9zdDo0",
            },
            {
              "$ref": "Post:UG9zdDo1",
            },
            {
              "$ref": "Post:UG9zdDo2",
            },
            {
              "$ref": "Post:UG9zdDo3",
            },
            {
              "$ref": "Post:UG9zdDo4",
            },
            {
              "$ref": "Post:UG9zdDo5",
            },
            {
              "$ref": "Post:UG9zdDoxMA==",
            },
          ],
        },
      },
      "{\\"first\\":\\"12\\"}": {
        "edges": {
          "$": [
            {
              "$ref": "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjow",
            },
            {
              "$ref": "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjox",
            },
            {
              "$ref": "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjoy",
            },
            {
              "$ref": "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjoz",
            },
            {
              "$ref": "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjo0",
            },
            {
              "$ref": "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjo1",
            },
            {
              "$ref": "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjo2",
            },
            {
              "$ref": "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjo3",
            },
            {
              "$ref": "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjo4",
            },
            {
              "$ref": "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjo5",
            },
            {
              "$ref": "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjoxMA==",
            },
            {
              "$ref": "QueryPostsConnectionEdge:T2Zmc2V0Q29ubmVjdGlvbjoxMQ==",
            },
          ],
        },
        "nodes": {
          "$": [
            {
              "$ref": "Post:UG9zdDox",
            },
            {
              "$ref": "Post:UG9zdDoy",
            },
            {
              "$ref": "Post:UG9zdDoz",
            },
            {
              "$ref": "Post:UG9zdDo0",
            },
            {
              "$ref": "Post:UG9zdDo1",
            },
            {
              "$ref": "Post:UG9zdDo2",
            },
            {
              "$ref": "Post:UG9zdDo3",
            },
            {
              "$ref": "Post:UG9zdDo4",
            },
            {
              "$ref": "Post:UG9zdDo5",
            },
            {
              "$ref": "Post:UG9zdDoxMA==",
            },
            {
              "$ref": "Post:UG9zdDoxMQ==",
            },
            {
              "$ref": "Post:UG9zdDoxMg==",
            },
          ],
        },
      },
    },
  },
  "modified": 0,
  "result": {
    "first10Posts": {
      "nodes": [
        {
          "id": "UG9zdDox",
          "title": "Some Blog post 1",
        },
        {
          "id": "UG9zdDoy",
          "title": "Some Blog post 2",
        },
        {
          "id": "UG9zdDoz",
          "title": "Some Blog post 3",
        },
        {
          "id": "UG9zdDo0",
          "title": "Some Blog post 4",
        },
        {
          "id": "UG9zdDo1",
          "title": "Some Blog post 5",
        },
        {
          "id": "UG9zdDo2",
          "title": "Some Blog post 6",
        },
        {
          "id": "UG9zdDo3",
          "title": "Some Blog post 7",
        },
        {
          "id": "UG9zdDo4",
          "title": "Some Blog post 8",
        },
        {
          "id": "UG9zdDo5",
          "title": "Some Blog post 9",
        },
        {
          "id": "UG9zdDoxMA==",
          "title": "Some Blog post 10",
        },
      ],
    },
    "homeItems": {
      "nodes": [
        {
          "__typename": "Comment",
        },
        {
          "id": "QmxvZzox",
          "name": "My Cool #1",
        },
        {
          "__typename": "Post",
          "id": "UG9zdDox",
        },
        {
          "__typename": "User",
        },
        {
          "__typename": "Comment",
        },
        {
          "id": "QmxvZzoy",
          "name": "My Cool #2",
        },
        {
          "__typename": "Post",
          "id": "UG9zdDoy",
        },
        {
          "__typename": "User",
        },
        {
          "__typename": "Comment",
        },
        {
          "id": "QmxvZzoz",
          "name": "My Cool #3",
        },
      ],
    },
    "nestedString": [
      [
        "a",
        "b",
      ],
      [
        "foo",
      ],
    ],
    "nestedUserList": [
      [
        {
          "name": "user 1",
        },
      ],
      [
        {
          "name": "user 2",
        },
      ],
    ],
    "num": 1,
    "posts": {
      "edges": [
        {
          "node": {
            "author": {
              "id": "VXNlcjox",
            },
            "id": "UG9zdDox",
          },
        },
        {
          "node": {
            "author": {
              "id": "VXNlcjoy",
            },
            "id": "UG9zdDoy",
          },
        },
        {
          "node": {
            "author": {
              "id": "VXNlcjox",
            },
            "id": "UG9zdDoz",
          },
        },
        {
          "node": {
            "author": {
              "id": "VXNlcjoy",
            },
            "id": "UG9zdDo0",
          },
        },
        {
          "node": {
            "author": {
              "id": "VXNlcjox",
            },
            "id": "UG9zdDo1",
          },
        },
        {
          "node": {
            "author": {
              "id": "VXNlcjoy",
            },
            "id": "UG9zdDo2",
          },
        },
        {
          "node": {
            "author": {
              "id": "VXNlcjox",
            },
            "id": "UG9zdDo3",
          },
        },
        {
          "node": {
            "author": {
              "id": "VXNlcjoy",
            },
            "id": "UG9zdDo4",
          },
        },
        {
          "node": {
            "author": {
              "id": "VXNlcjox",
            },
            "id": "UG9zdDo5",
          },
        },
        {
          "node": {
            "author": {
              "id": "VXNlcjoy",
            },
            "id": "UG9zdDoxMA==",
          },
        },
        {
          "node": {
            "author": {
              "id": "VXNlcjox",
            },
            "id": "UG9zdDoxMQ==",
          },
        },
        {
          "node": {
            "author": {
              "id": "VXNlcjoy",
            },
            "id": "UG9zdDoxMg==",
          },
        },
      ],
      "nodes": [
        {
          "author": {
            "id": "VXNlcjox",
          },
          "id": "UG9zdDox",
        },
        {
          "author": {
            "id": "VXNlcjoy",
          },
          "id": "UG9zdDoy",
        },
        {
          "author": {
            "id": "VXNlcjox",
          },
          "id": "UG9zdDoz",
        },
        {
          "author": {
            "id": "VXNlcjoy",
          },
          "id": "UG9zdDo0",
        },
        {
          "author": {
            "id": "VXNlcjox",
          },
          "id": "UG9zdDo1",
        },
        {
          "author": {
            "id": "VXNlcjoy",
          },
          "id": "UG9zdDo2",
        },
        {
          "author": {
            "id": "VXNlcjox",
          },
          "id": "UG9zdDo3",
        },
        {
          "author": {
            "id": "VXNlcjoy",
          },
          "id": "UG9zdDo4",
        },
        {
          "author": {
            "id": "VXNlcjox",
          },
          "id": "UG9zdDo5",
        },
        {
          "author": {
            "id": "VXNlcjoy",
          },
          "id": "UG9zdDoxMA==",
        },
        {
          "author": {
            "id": "VXNlcjox",
          },
          "id": "UG9zdDoxMQ==",
        },
        {
          "author": {
            "id": "VXNlcjoy",
          },
          "id": "UG9zdDoxMg==",
        },
      ],
    },
  },
}
`;

exports[`syncWithCache > writes additional values into the store when the variables are changed 1`] = `
[
  {
    "op": "add",
    "path": [
      "cache",
      "node",
    ],
    "value": {
      "{\\"id\\":\\"VXNlcjox\\"}": {
        "$ref": "User:VXNlcjox",
      },
    },
  },
  {
    "op": "add",
    "path": [
      "currentResult",
      "currentNode",
    ],
    "value": {
      "__typename": "User",
      "name": "user 1",
    },
  },
]
`;
